# Más sobre la libreria `pandas`

La información vienen comunmente en formatos excel, csv, txt, o quizá esté en alguna url. Aunque estos pueden estar en otro formato (JSON, API'S) nos centraremos en los tres primeros.

In [1]:
ls -l

total 2236
drwxrwxr-x 5 andres andres   4096 jun 30 00:11 _build/
-rw-rw-r-- 1 andres andres   9121 jun 30 00:48 Clases_Python.ipynb
-rw-rw-r-- 1 andres andres   1006 jun 30 00:48 _config.yml
drwxrwxr-x 5 andres andres   4096 jun 30 11:44 datas/
-rw-rw-r-- 1 andres andres      0 jun 30 00:48 hello.txt
-rw-rw-r-- 1 andres andres    754 jun 30 00:48 intro.md
-rw-rw-r-- 1 andres andres  12653 jun 30 00:48 Lectura_1_Intro_Markdown.ipynb
-rw-rw-r-- 1 andres andres  74974 jun 30 00:48 Lectura_2_elementos_basicos_Python.ipynb
-rw-rw-r-- 1 andres andres 995211 jun 30 00:48 Lectura_3_Defi_Function_Python.ipynb
-rw-rw-r-- 1 andres andres  74710 jun 30 11:45 Lectura_4_estruc_Datos_Python.ipynb
-rw-rw-r-- 1 andres andres  48871 jun 30 11:33 Lectura_5_Pandas_Series_DataFrame.ipynb
-rw-rw-r-- 1 andres andres 198309 jun 30 12:28 Lectura_6.0_Mas_de_Pandas.ipynb
-rw-rw-r-- 1 andres andres 733688 jun 30 12:21 Lectura_6.1_operacionesBásicasDataframes.ipynb
-rw-rw-r-- 1 andres andres  83564 jun 30 00:48 l

In [2]:
## Librerías requeridas
import numpy as np
import pandas as pd

## pd.set_option('display.max_columns', None) ## Número de columnas máximo a mostrar. 
                                              ## None se reemplaza por el # máximo de columnas.

In [3]:
## Función .head() muestra los 5 primeros elementos
df_csv = pd.read_csv('datas/downloaded_medals.csv', sep = ",")
df_csv.head()

Unnamed: 0      City      Discipline       Event Event gender   Medal  NOC  \
0           0  Chamonix  Figure skating  individual            M  Silver  AUT   
1           1  Chamonix  Figure skating  individual            W    Gold  AUT   
2           2  Chamonix  Figure skating       pairs            X    Gold  AUT   
3           3  Chamonix       Bobsleigh    four-man            M  Bronze  BEL   
4           4  Chamonix      Ice Hockey  ice hockey            M    Gold  CAN   

        Sport  Year  
0     Skating  1924  
1     Skating  1924  
2     Skating  1924  
3   Bobsleigh  1924  
4  Ice Hockey  1924

In [4]:
df_csv = df_csv.drop(columns=['Unnamed: 0']) ## Borrar columna
df_csv.head()

City      Discipline       Event Event gender   Medal  NOC       Sport  \
0  Chamonix  Figure skating  individual            M  Silver  AUT     Skating   
1  Chamonix  Figure skating  individual            W    Gold  AUT     Skating   
2  Chamonix  Figure skating       pairs            X    Gold  AUT     Skating   
3  Chamonix       Bobsleigh    four-man            M  Bronze  BEL   Bobsleigh   
4  Chamonix      Ice Hockey  ice hockey            M    Gold  CAN  Ice Hockey   

   Year  
0  1924  
1  1924  
2  1924  
3  1924  
4  1924

In [5]:
## Función .tail() muestra los 5 últimos elementos
df_csv = pd.read_csv('datas/downloaded_medals.csv', index_col=0)
df_csv.tail()

City Discipline            Event Event gender   Medal  NOC   Sport  \
2306  Turin  Snowboard        Half-pipe            M  Silver  USA  Skiing   
2307  Turin  Snowboard        Half-pipe            W    Gold  USA  Skiing   
2308  Turin  Snowboard        Half-pipe            W  Silver  USA  Skiing   
2309  Turin  Snowboard  Snowboard Cross            M    Gold  USA  Skiing   
2310  Turin  Snowboard  Snowboard Cross            W  Silver  USA  Skiing   

      Year  
2306  2006  
2307  2006  
2308  2006  
2309  2006  
2310  2006

## _Archivos excel._

In [6]:
# Archivos Excel
df_Excel= pd.ExcelFile('datas/downloaded_medals.xls')
df_Excel.sheet_names

['Sheet1']

In [7]:
df_Excel = df_Excel.parse("Sheet1")
df_Excel.head()

Unnamed: 0      City      Discipline       Event Event gender   Medal  NOC  \
0           0  Chamonix  Figure skating  individual            M  Silver  AUT   
1           1  Chamonix  Figure skating  individual            W    Gold  AUT   
2           2  Chamonix  Figure skating       pairs            X    Gold  AUT   
3           3  Chamonix       Bobsleigh    four-man            M  Bronze  BEL   
4           4  Chamonix      Ice Hockey  ice hockey            M    Gold  CAN   

        Sport  Year  
0     Skating  1924  
1     Skating  1924  
2     Skating  1924  
3   Bobsleigh  1924  
4  Ice Hockey  1924

# Archivos JSON.

In [8]:
df_json = pd.read_json("datas/downloaded_medals.json")
df_json.head()

City      Discipline       Event Event gender   Medal  NOC       Sport  \
0  Chamonix  Figure skating  individual            M  Silver  AUT     Skating   
1  Chamonix  Figure skating  individual            W    Gold  AUT     Skating   
2  Chamonix  Figure skating       pairs            X    Gold  AUT     Skating   
3  Chamonix       Bobsleigh    four-man            M  Bronze  BEL   Bobsleigh   
4  Chamonix      Ice Hockey  ice hockey            M    Gold  CAN  Ice Hockey   

   Year  
0  1924  
1  1924  
2  1924  
3  1924  
4  1924

#### Ejercicio 1

Importar el archivo `'Winter Olympic Medals.xlsx'` y definir en cada caso un dataframe para cada hoja. Además en cada caso, mostrar su cabeza y su cola.

**Respuesta:**

In [9]:
#celda de código para probar.
df_Excel_1 = pd.ExcelFile('datas/Winter Olympic Medals.xlsx')
df_Excel_1.sheet_names

['Data', 'Ranks', 'Sources']

In [10]:
df_excel = pd.read_excel('datas/Winter Olympic Medals.xlsx', sheet_name=['Data', 'Ranks', 'Sources'])
df_aux = df_excel["Data"]
df_aux.head()

Year                 Sport                Event      Country Gender  \
0  1924               Bobsled      Men's Four/Five  Switzerland    Men   
1  1924               Bobsled      Men's Four/Five      Britain    Men   
2  1924               Bobsled      Men's Four/Five      Belgium    Men   
3  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   

   Medal Rank   Medal Name of Athlete or Team  Age of Athlete  
0           1    gold           Switzerland-1             NaN  
1           2  silver               Britain-1             NaN  
2           3  bronze               Belgium-1             NaN  
3           1    gold           Thorleif Haug            29.0  
4           2  silver  Johan Grøttumsbraaten            24.0

## _Desde una URL._


In [11]:
## Dependerá de donde y como estén los datos.
df_url = pd.read_excel('https://query.data.world/s/zilmgkqzxjhjrfn2pllc4dhikcvrzt')
df_url.head()

Year                 Sport                Event      Country Gender  \
0  1924               Bobsled      Men's Four/Five  Switzerland    Men   
1  1924               Bobsled      Men's Four/Five      Britain    Men   
2  1924               Bobsled      Men's Four/Five      Belgium    Men   
3  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   

   Medal Rank   Medal Name of Athlete or Team  Age of Athlete  
0           1    gold           Switzerland-1             NaN  
1           2  silver               Britain-1             NaN  
2           3  bronze               Belgium-1             NaN  
3           1    gold           Thorleif Haug            29.0  
4           2  silver  Johan Grøttumsbraaten            24.0

In [12]:
url = "http://winterolympicsmedals.com/medals.csv"
df_url_1 = pd.read_csv(url)
df_url_1.head()

Year      City       Sport      Discipline  NOC       Event Event gender  \
0  1924  Chamonix     Skating  Figure skating  AUT  individual            M   
1  1924  Chamonix     Skating  Figure skating  AUT  individual            W   
2  1924  Chamonix     Skating  Figure skating  AUT       pairs            X   
3  1924  Chamonix   Bobsleigh       Bobsleigh  BEL    four-man            M   
4  1924  Chamonix  Ice Hockey      Ice Hockey  CAN  ice hockey            M   

    Medal  
0  Silver  
1    Gold  
2    Gold  
3  Bronze  
4    Gold

## _Información del `dataframe`_


In [13]:
# Copiamos un dataframe
df = df_url.copy()
df.head()

Year                 Sport                Event      Country Gender  \
0  1924               Bobsled      Men's Four/Five  Switzerland    Men   
1  1924               Bobsled      Men's Four/Five      Britain    Men   
2  1924               Bobsled      Men's Four/Five      Belgium    Men   
3  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   

   Medal Rank   Medal Name of Athlete or Team  Age of Athlete  
0           1    gold           Switzerland-1             NaN  
1           2  silver               Britain-1             NaN  
2           3  bronze               Belgium-1             NaN  
3           1    gold           Thorleif Haug            29.0  
4           2  silver  Johan Grøttumsbraaten            24.0

In [14]:
## Nombres columnas
df.columns

Index(['Year', 'Sport', 'Event', 'Country', 'Gender', 'Medal Rank', 'Medal',
       'Name of Athlete or Team', 'Age of Athlete'],
      dtype='object')

In [15]:
## Tipos de datos en el dataframe
df.dtypes

Year                         int64
Sport                       object
Event                       object
Country                     object
Gender                      object
Medal Rank                   int64
Medal                       object
Name of Athlete or Team     object
Age of Athlete             float64
dtype: object

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2865 entries, 0 to 2864
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Year                     2865 non-null   int64  
 1   Sport                    2865 non-null   object 
 2   Event                    2865 non-null   object 
 3   Country                  2865 non-null   object 
 4   Gender                   2865 non-null   object 
 5   Medal Rank               2865 non-null   int64  
 6   Medal                    2865 non-null   object 
 7   Name of Athlete or Team  2865 non-null   object 
 8   Age of Athlete           2173 non-null   float64
dtypes: float64(1), int64(2), object(6)
memory usage: 201.6+ KB


In [17]:
df.describe().T # df.describe.transpose()

count         mean        std     min     25%     50%  \
Year            2865.0  1986.488656  23.489775  1924.0  1972.0  1992.0   
Medal Rank      2865.0     1.995812   0.815773     1.0     1.0     2.0   
Age of Athlete  2173.0    25.152324   4.271764    14.0    22.0    25.0   

                   75%     max  
Year            2006.0  2014.0  
Medal Rank         3.0     3.0  
Age of Athlete    28.0    42.0

In [18]:
df = df.rename(columns={"Year":"Año", "Sport":"Deporte", "Country":"País", "Event":"Evento","Gender":"Género","Medal":"Medalla",
                        "Age of Athlete":"Edad_Atleta","Name of Athlete or Team":"Nombre_atleta_o_Equipo"})
df.head()

Año               Deporte               Evento         País Género  \
0  1924               Bobsled      Men's Four/Five  Switzerland    Men   
1  1924               Bobsled      Men's Four/Five      Britain    Men   
2  1924               Bobsled      Men's Four/Five      Belgium    Men   
3  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   

   Medal Rank Medalla  Nombre_atleta_o_Equipo  Edad_Atleta  
0           1    gold           Switzerland-1          NaN  
1           2  silver               Britain-1          NaN  
2           3  bronze               Belgium-1          NaN  
3           1    gold           Thorleif Haug         29.0  
4           2  silver  Johan Grøttumsbraaten         24.0

### _Sub dataframes._


In [19]:
## Serie de pandas
df['Edad_Atleta']

0        NaN
1        NaN
2        NaN
3       29.0
4       24.0
        ... 
2860    30.0
2861    31.0
2862    27.0
2863     NaN
2864     NaN
Name: Edad_Atleta, Length: 2865, dtype: float64

In [20]:
df['Evento']

0           Men's Four/Five
1           Men's Four/Five
2           Men's Four/Five
3       Men's 18 Kilometers
4       Men's 18 Kilometers
               ...         
2860    Men's 50 Kilometers
2861    Men's 50 Kilometers
2862    Men's 50 Kilometers
2863       Men's Ice Hockey
2864       Men's Ice Hockey
Name: Evento, Length: 2865, dtype: object

In [21]:
df['Género'].unique()

array(['Men', 'Mixed', 'Women'], dtype=object)

In [22]:
subset = df[['Año', 'Deporte', 'Evento']]
subset

Año               Deporte               Evento
0     1924               Bobsled      Men's Four/Five
1     1924               Bobsled      Men's Four/Five
2     1924               Bobsled      Men's Four/Five
3     1924  Cross-Country Skiing  Men's 18 Kilometers
4     1924  Cross-Country Skiing  Men's 18 Kilometers
...    ...                   ...                  ...
2860  2014  Cross-Country Skiing  Men's 50 Kilometers
2861  2014  Cross-Country Skiing  Men's 50 Kilometers
2862  2014  Cross-Country Skiing  Men's 50 Kilometers
2863  2014            Ice Hockey     Men's Ice Hockey
2864  2014            Ice Hockey     Men's Ice Hockey

[2865 rows x 3 columns]

In [23]:
df["Deporte"].value_counts()

Speedskating                527
Cross-Country Skiing        471
Alpine Skiing               428
Figure Skating              235
Biathlon                    225
Short-Track Speedskating    144
Ski Jumping                 135
Bobsled                     133
Luge                        129
Nordic Combined             102
Freestyle Skiing            102
Snowboarding                 90
Ice Hockey                   81
Curling                      33
Skeleton                     30
Name: Deporte, dtype: int64

In [24]:
df["Deporte"].value_counts(1,dropna=False)

Speedskating                0.183944
Cross-Country Skiing        0.164398
Alpine Skiing               0.149389
Figure Skating              0.082024
Biathlon                    0.078534
Short-Track Speedskating    0.050262
Ski Jumping                 0.047120
Bobsled                     0.046422
Luge                        0.045026
Nordic Combined             0.035602
Freestyle Skiing            0.035602
Snowboarding                0.031414
Ice Hockey                  0.028272
Curling                     0.011518
Skeleton                    0.010471
Name: Deporte, dtype: float64

In [25]:
df['Género'].value_counts(1,dropna=False)*100

Men      59.720768
Women    36.474695
Mixed     3.804538
Name: Género, dtype: float64

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2865 entries, 0 to 2864
Data columns (total 9 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Año                     2865 non-null   int64  
 1   Deporte                 2865 non-null   object 
 2   Evento                  2865 non-null   object 
 3   País                    2865 non-null   object 
 4   Género                  2865 non-null   object 
 5   Medal Rank              2865 non-null   int64  
 6   Medalla                 2865 non-null   object 
 7   Nombre_atleta_o_Equipo  2865 non-null   object 
 8   Edad_Atleta             2173 non-null   float64
dtypes: float64(1), int64(2), object(6)
memory usage: 201.6+ KB


In [27]:
numeric = ["Edad_Atleta"]
other_features = [x for x in df.columns if x not in numeric] ## list comprehension
other_features

['Año',
 'Deporte',
 'Evento',
 'País',
 'Género',
 'Medal Rank',
 'Medalla',
 'Nombre_atleta_o_Equipo']

In [28]:
df_categoric = df[other_features]
df_categoric.head()

Año               Deporte               Evento         País Género  \
0  1924               Bobsled      Men's Four/Five  Switzerland    Men   
1  1924               Bobsled      Men's Four/Five      Britain    Men   
2  1924               Bobsled      Men's Four/Five      Belgium    Men   
3  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   

   Medal Rank Medalla  Nombre_atleta_o_Equipo  
0           1    gold           Switzerland-1  
1           2  silver               Britain-1  
2           3  bronze               Belgium-1  
3           1    gold           Thorleif Haug  
4           2  silver  Johan Grøttumsbraaten

## `.loc` y `.iloc`


Si deseamos seleccionar un subdataframe, con registros y columnas particulares podemos hacerlos con estos 2 métodos. `.loc` es la localización en texto (elas columnas) y `.iloc()` será vía indexación numérica.

In [29]:
df.head()

Año               Deporte               Evento         País Género  \
0  1924               Bobsled      Men's Four/Five  Switzerland    Men   
1  1924               Bobsled      Men's Four/Five      Britain    Men   
2  1924               Bobsled      Men's Four/Five      Belgium    Men   
3  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   

   Medal Rank Medalla  Nombre_atleta_o_Equipo  Edad_Atleta  
0           1    gold           Switzerland-1          NaN  
1           2  silver               Britain-1          NaN  
2           3  bronze               Belgium-1          NaN  
3           1    gold           Thorleif Haug         29.0  
4           2  silver  Johan Grøttumsbraaten         24.0

In [30]:
df.loc[4] ## Registro 4 de la tabla.

Año                                         1924
Deporte                     Cross-Country Skiing
Evento                       Men's 18 Kilometers
País                                      Norway
Género                                       Men
Medal Rank                                     2
Medalla                                   silver
Nombre_atleta_o_Equipo    Johan Grøttumsbraaten
Edad_Atleta                                 24.0
Name: 4, dtype: object

In [31]:
type(df.loc[4])

pandas.core.series.Series

In [32]:
df.loc[:8] ## 8 Primeros registros del dataframe

Año               Deporte               Evento         País Género  \
0  1924               Bobsled      Men's Four/Five  Switzerland    Men   
1  1924               Bobsled      Men's Four/Five      Britain    Men   
2  1924               Bobsled      Men's Four/Five      Belgium    Men   
3  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
5  1924  Cross-Country Skiing  Men's 18 Kilometers      Finland    Men   
6  1924  Cross-Country Skiing  Men's 50 Kilometers       Norway    Men   
7  1924  Cross-Country Skiing  Men's 50 Kilometers       Norway    Men   
8  1924  Cross-Country Skiing  Men's 50 Kilometers       Norway    Men   

   Medal Rank Medalla  Nombre_atleta_o_Equipo  Edad_Atleta  
0           1    gold           Switzerland-1          NaN  
1           2  silver               Britain-1          NaN  
2           3  bronze               Belgium-1          NaN  
3           1    gold           Thorleif Haug         29.0  
4           2  silver  Johan Grøttumsbraaten         24.0  
5           3  bronze             Tapani Niku         28.0  
6           1    gold           Thorleif Haug         29.0  
7           2  silver      Thoralf Strømstad         27.0  
8           3  bronze  Johan Grøttumsbraaten         24.0

In [33]:
df.iloc[10:20] # Registros desde el 11 y hasta el 20

Año         Deporte           Evento           País Género  Medal Rank  \
10  1924         Curling    Men's Curling         Sweden    Men           2   
11  1924         Curling    Men's Curling         France    Men           3   
12  1924  Figure Skating    Men's Singles         Sweden    Men           1   
13  1924  Figure Skating    Men's Singles        Austria    Men           2   
14  1924  Figure Skating    Men's Singles    Switzerland    Men           3   
15  1924  Figure Skating      Mixed Pairs        Austria  Mixed           1   
16  1924  Figure Skating      Mixed Pairs        Finland  Mixed           2   
17  1924  Figure Skating      Mixed Pairs         France  Mixed           3   
18  1924  Figure Skating  Women's Singles        Austria  Women           1   
19  1924  Figure Skating  Women's Singles  United States  Women           2   

   Medalla Nombre_atleta_o_Equipo  Edad_Atleta  
10  silver                 Sweden          NaN  
11  bronze                 France          NaN  
12    gold      Gillis Grafström         30.0  
13  silver           Willy Böckl         30.0  
14  bronze       Georges Gautschi         19.0  
15    gold                Austria          NaN  
16  silver                Finland          NaN  
17  bronze               France-1          NaN  
18    gold     Herma Planck-Szabo         21.0  
19  silver       Beatrix Loughran         23.0

In [34]:
pares = [i for i in range(len(df['Año'])) if i%2==0]
df.loc[pares]

Año               Deporte               Evento         País Género  \
0     1924               Bobsled      Men's Four/Five  Switzerland    Men   
2     1924               Bobsled      Men's Four/Five      Belgium    Men   
4     1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
6     1924  Cross-Country Skiing  Men's 50 Kilometers       Norway    Men   
8     1924  Cross-Country Skiing  Men's 50 Kilometers       Norway    Men   
...    ...                   ...                  ...          ...    ...   
2856  2014          Speedskating   Men's Team Pursuit       Poland    Men   
2858  2014               Bobsled           Men's Four       Latvia    Men   
2860  2014  Cross-Country Skiing  Men's 50 Kilometers       Russia    Men   
2862  2014  Cross-Country Skiing  Men's 50 Kilometers       Russia    Men   
2864  2014            Ice Hockey     Men's Ice Hockey       Sweden    Men   

      Medal Rank Medalla  Nombre_atleta_o_Equipo  Edad_Atleta  
0              1    gold           Switzerland-1          NaN  
2              3  bronze               Belgium-1          NaN  
4              2  silver  Johan Grøttumsbraaten         24.0  
6              1    gold           Thorleif Haug         29.0  
8              3  bronze  Johan Grøttumsbraaten         24.0  
...          ...     ...                     ...          ...  
2856           3  bronze                  Poland          NaN  
2858           2  silver                  Latvia          NaN  
2860           1    gold        Alexander Legkov         30.0  
2862           3  bronze         Ilia Chernousov         27.0  
2864           2  silver                  Sweden          NaN  

[1433 rows x 9 columns]

In [35]:
impares = [i for i in range(len(df['Año'])) if i%2==1]
df.iloc[impares]

Año               Deporte               Evento           País Género  \
1     1924               Bobsled      Men's Four/Five        Britain    Men   
3     1924  Cross-Country Skiing  Men's 18 Kilometers         Norway    Men   
5     1924  Cross-Country Skiing  Men's 18 Kilometers        Finland    Men   
7     1924  Cross-Country Skiing  Men's 50 Kilometers         Norway    Men   
9     1924               Curling        Men's Curling        Britain    Men   
...    ...                   ...                  ...            ...    ...   
2855  2014          Speedskating   Men's Team Pursuit    South Korea    Men   
2857  2014               Bobsled           Men's Four         Russia    Men   
2859  2014               Bobsled           Men's Four  United States    Men   
2861  2014  Cross-Country Skiing  Men's 50 Kilometers         Russia    Men   
2863  2014            Ice Hockey     Men's Ice Hockey         Canada    Men   

      Medal Rank Medalla Nombre_atleta_o_Equipo  Edad_Atleta  
1              2  silver              Britain-1          NaN  
3              1    gold          Thorleif Haug         29.0  
5              3  bronze            Tapani Niku         28.0  
7              2  silver     Thoralf Strømstad         27.0  
9              1    gold                Britain          NaN  
...          ...     ...                    ...          ...  
2855           2  silver            South Korea          NaN  
2857           1    gold                 Russia          NaN  
2859           3  bronze          United States          NaN  
2861           2  silver      Maxim Vylegzhanin         31.0  
2863           1    gold                 Canada          NaN  

[1432 rows x 9 columns]

In [36]:
pd.DataFrame(df.loc[:,"Deporte"])

Deporte
0                  Bobsled
1                  Bobsled
2                  Bobsled
3     Cross-Country Skiing
4     Cross-Country Skiing
...                    ...
2860  Cross-Country Skiing
2861  Cross-Country Skiing
2862  Cross-Country Skiing
2863            Ice Hockey
2864            Ice Hockey

[2865 rows x 1 columns]

In [37]:
df.loc[:,["Deporte","Evento"]]

Deporte               Evento
0                  Bobsled      Men's Four/Five
1                  Bobsled      Men's Four/Five
2                  Bobsled      Men's Four/Five
3     Cross-Country Skiing  Men's 18 Kilometers
4     Cross-Country Skiing  Men's 18 Kilometers
...                    ...                  ...
2860  Cross-Country Skiing  Men's 50 Kilometers
2861  Cross-Country Skiing  Men's 50 Kilometers
2862  Cross-Country Skiing  Men's 50 Kilometers
2863            Ice Hockey     Men's Ice Hockey
2864            Ice Hockey     Men's Ice Hockey

[2865 rows x 2 columns]

In [38]:
for i,j in enumerate(df.columns):
  print((i,j))

(0, 'Año')
(1, 'Deporte')
(2, 'Evento')
(3, 'País')
(4, 'Género')
(5, 'Medal Rank')
(6, 'Medalla')
(7, 'Nombre_atleta_o_Equipo')
(8, 'Edad_Atleta')


In [39]:
df.iloc[pares,[0,1,7]]

Año               Deporte  Nombre_atleta_o_Equipo
0     1924               Bobsled           Switzerland-1
2     1924               Bobsled               Belgium-1
4     1924  Cross-Country Skiing  Johan Grøttumsbraaten
6     1924  Cross-Country Skiing           Thorleif Haug
8     1924  Cross-Country Skiing  Johan Grøttumsbraaten
...    ...                   ...                     ...
2856  2014          Speedskating                  Poland
2858  2014               Bobsled                  Latvia
2860  2014  Cross-Country Skiing        Alexander Legkov
2862  2014  Cross-Country Skiing         Ilia Chernousov
2864  2014            Ice Hockey                  Sweden

[1433 rows x 3 columns]

## _Máscaras._

Supongamos que deseamos obtener los regístros que cumplen cierta condición. Por ejemplo, todos aquellos donde el género sea mujer, las personas menores de 30 años, aquellos que ganaron plata, o quizá mujeres que ganaron plata y son menores de 30 años.

Para hacer esto necesitamos una serie de `pandas` de tipo booleano, que determina que registros cumplen la condición deseada.

In [40]:
df.head()

Año               Deporte               Evento         País Género  \
0  1924               Bobsled      Men's Four/Five  Switzerland    Men   
1  1924               Bobsled      Men's Four/Five      Britain    Men   
2  1924               Bobsled      Men's Four/Five      Belgium    Men   
3  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   

   Medal Rank Medalla  Nombre_atleta_o_Equipo  Edad_Atleta  
0           1    gold           Switzerland-1          NaN  
1           2  silver               Britain-1          NaN  
2           3  bronze               Belgium-1          NaN  
3           1    gold           Thorleif Haug         29.0  
4           2  silver  Johan Grøttumsbraaten         24.0

In [41]:
mask = (df['Año'] > 1950)
mask

0       False
1       False
2       False
3       False
4       False
        ...  
2860     True
2861     True
2862     True
2863     True
2864     True
Name: Año, Length: 2865, dtype: bool

In [42]:
## Tenemos todos los registros de quienes ganaron medalla después del Año 1948
df[mask]

Año               Deporte               Evento     País Género  \
251   1952         Alpine Skiing       Men's Downhill    Italy    Men   
252   1952         Alpine Skiing       Men's Downhill  Austria    Men   
253   1952         Alpine Skiing       Men's Downhill  Austria    Men   
254   1952         Alpine Skiing   Men's Giant Slalom   Norway    Men   
255   1952         Alpine Skiing   Men's Giant Slalom  Austria    Men   
...    ...                   ...                  ...      ...    ...   
2860  2014  Cross-Country Skiing  Men's 50 Kilometers   Russia    Men   
2861  2014  Cross-Country Skiing  Men's 50 Kilometers   Russia    Men   
2862  2014  Cross-Country Skiing  Men's 50 Kilometers   Russia    Men   
2863  2014            Ice Hockey     Men's Ice Hockey   Canada    Men   
2864  2014            Ice Hockey     Men's Ice Hockey   Sweden    Men   

      Medal Rank Medalla Nombre_atleta_o_Equipo  Edad_Atleta  
251            1    gold             Zeno Colò         31.0  
252            2  silver       Othmar Schneider         23.0  
253            3  bronze       Christian Pravda         24.0  
254            1    gold          Stein Eriksen         24.0  
255            2  silver       Christian Pravda         24.0  
...          ...     ...                    ...          ...  
2860           1    gold       Alexander Legkov         30.0  
2861           2  silver      Maxim Vylegzhanin         31.0  
2862           3  bronze        Ilia Chernousov         27.0  
2863           1    gold                 Canada          NaN  
2864           2  silver                 Sweden          NaN  

[2614 rows x 9 columns]

In [43]:
mask_1 = (df['Edad_Atleta'] <= 20)

In [44]:
menores_de_20 = df[mask_1]
menores_de_20

Año                   Deporte                Evento           País  \
14    1924            Figure Skating         Men's Singles    Switzerland   
60    1928            Figure Skating         Men's Singles        Belgium   
64    1928            Figure Skating       Women's Singles         Norway   
65    1928            Figure Skating       Women's Singles        Austria   
74    1928                  Skeleton        Men's Skeleton  United States   
...    ...                       ...                   ...            ...   
2814  2014             Alpine Skiing        Women's Slalom  United States   
2827  2014  Short-Track Speedskating  Women's 1,000 Meters          China   
2828  2014  Short-Track Speedskating  Women's 1,000 Meters    South Korea   
2830  2014  Short-Track Speedskating      Men's 500 Meters          China   
2837  2014             Alpine Skiing          Men's Slalom         Norway   

     Género  Medal Rank Medalla Nombre_atleta_o_Equipo  Edad_Atleta  
14      Men           3  bronze       Georges Gautschi         19.0  
60      Men           3  bronze   Robert Van Zeebroeck         18.0  
64    Women           1    gold            Sonja Henie         15.0  
65    Women           2  silver          Fritzi Burger         17.0  
74      Men           2  silver            Jack Heaton         19.0  
...     ...         ...     ...                    ...          ...  
2814  Women           1    gold       Mikaela Shiffrin         18.0  
2827  Women           2  silver              Kexin Fan         20.0  
2828  Women           3  bronze           Suk Hee Shim         17.0  
2830    Men           2  silver              Dajing Wu         19.0  
2837    Men           3  bronze   Henrik Kristoffersen         19.0  

[299 rows x 9 columns]

In [45]:
menores_de_20.describe().T

count         mean        std     min     25%     50%     75%  \
Año          299.0  1983.471572  22.087115  1924.0  1968.0  1984.0  2002.0   
Medal Rank   299.0     2.023411   0.812726     1.0     1.0     2.0     3.0   
Edad_Atleta  299.0    18.856187   1.332123    14.0    18.0    19.0    20.0   

                max  
Año          2014.0  
Medal Rank      3.0  
Edad_Atleta    20.0

In [46]:
menores_de_20.reset_index()

index   Año                   Deporte                Evento  \
0       14  1924            Figure Skating         Men's Singles   
1       60  1928            Figure Skating         Men's Singles   
2       64  1928            Figure Skating       Women's Singles   
3       65  1928            Figure Skating       Women's Singles   
4       74  1928                  Skeleton        Men's Skeleton   
..     ...   ...                       ...                   ...   
294   2814  2014             Alpine Skiing        Women's Slalom   
295   2827  2014  Short-Track Speedskating  Women's 1,000 Meters   
296   2828  2014  Short-Track Speedskating  Women's 1,000 Meters   
297   2830  2014  Short-Track Speedskating      Men's 500 Meters   
298   2837  2014             Alpine Skiing          Men's Slalom   

              País Género  Medal Rank Medalla Nombre_atleta_o_Equipo  \
0      Switzerland    Men           3  bronze       Georges Gautschi   
1          Belgium    Men           3  bronze   Robert Van Zeebroeck   
2           Norway  Women           1    gold            Sonja Henie   
3          Austria  Women           2  silver          Fritzi Burger   
4    United States    Men           2  silver            Jack Heaton   
..             ...    ...         ...     ...                    ...   
294  United States  Women           1    gold       Mikaela Shiffrin   
295          China  Women           2  silver              Kexin Fan   
296    South Korea  Women           3  bronze           Suk Hee Shim   
297          China    Men           2  silver              Dajing Wu   
298         Norway    Men           3  bronze   Henrik Kristoffersen   

     Edad_Atleta  
0           19.0  
1           18.0  
2           15.0  
3           17.0  
4           19.0  
..           ...  
294         18.0  
295         20.0  
296         17.0  
297         19.0  
298         19.0  

[299 rows x 10 columns]

In [47]:
menores_de_20.reset_index(drop=True,)

Año                   Deporte                Evento           País  \
0    1924            Figure Skating         Men's Singles    Switzerland   
1    1928            Figure Skating         Men's Singles        Belgium   
2    1928            Figure Skating       Women's Singles         Norway   
3    1928            Figure Skating       Women's Singles        Austria   
4    1928                  Skeleton        Men's Skeleton  United States   
..    ...                       ...                   ...            ...   
294  2014             Alpine Skiing        Women's Slalom  United States   
295  2014  Short-Track Speedskating  Women's 1,000 Meters          China   
296  2014  Short-Track Speedskating  Women's 1,000 Meters    South Korea   
297  2014  Short-Track Speedskating      Men's 500 Meters          China   
298  2014             Alpine Skiing          Men's Slalom         Norway   

    Género  Medal Rank Medalla Nombre_atleta_o_Equipo  Edad_Atleta  
0      Men           3  bronze       Georges Gautschi         19.0  
1      Men           3  bronze   Robert Van Zeebroeck         18.0  
2    Women           1    gold            Sonja Henie         15.0  
3    Women           2  silver          Fritzi Burger         17.0  
4      Men           2  silver            Jack Heaton         19.0  
..     ...         ...     ...                    ...          ...  
294  Women           1    gold       Mikaela Shiffrin         18.0  
295  Women           2  silver              Kexin Fan         20.0  
296  Women           3  bronze           Suk Hee Shim         17.0  
297    Men           2  silver              Dajing Wu         19.0  
298    Men           3  bronze   Henrik Kristoffersen         19.0  

[299 rows x 9 columns]

In [48]:
## Podemos juntar ambas condiciones
menores_ano_cincuenta = df[mask&mask_1]
menores_ano_cincuenta = menores_ano_cincuenta.reset_index(drop=True)
menores_ano_cincuenta.head()

Año         Deporte                Evento           País Género  \
0  1952   Alpine Skiing      Women's Downhill          Italy  Women   
1  1952   Alpine Skiing  Women's Giant Slalom  United States  Women   
2  1952   Alpine Skiing        Women's Slalom  United States  Women   
3  1952  Figure Skating         Men's Singles  United States    Men   
4  1952  Figure Skating       Women's Singles  United States  Women   

   Medal Rank Medalla   Nombre_atleta_o_Equipo  Edad_Atleta  
0           3  bronze  Giuliana Chenal-Minuzzo         20.0  
1           1    gold     Andrea Mead-Lawrence         19.0  
2           1    gold     Andrea Mead-Lawrence         19.0  
3           3  bronze               Jim Grogan         20.0  
4           2  silver          Tenley Albright         16.0

In [49]:
menores_ano_cincuenta["Medalla"].value_counts()

bronze    94
silver    93
gold      89
Name: Medalla, dtype: int64

In [50]:
df[mask & mask_1]["Medalla"].value_counts()

bronze    94
silver    93
gold      89
Name: Medalla, dtype: int64

In [51]:
mask_2 = df['País'] == "China"
df[mask & mask_1 & mask_2]#["Género"].value_counts()*100

Año                   Deporte                Evento   País Género  \
1531  1994            Figure Skating       Women's Singles  China  Women   
1732  1998          Freestyle Skiing       Women's Aerials  China  Women   
1756  1998  Short-Track Speedskating      Men's 500 Meters  China    Men   
1765  1998  Short-Track Speedskating    Women's 500 Meters  China  Women   
1768  1998  Short-Track Speedskating  Women's 1,000 Meters  China  Women   
2229  2006  Short-Track Speedskating    Women's 500 Meters  China  Women   
2233  2006  Short-Track Speedskating  Women's 1,000 Meters  China  Women   
2237  2006  Short-Track Speedskating  Women's 1,500 Meters  China  Women   
2493  2010  Short-Track Speedskating  Women's 1,500 Meters  China  Women   
2616  2014  Short-Track Speedskating    Men's 1,500 Meters  China    Men   
2827  2014  Short-Track Speedskating  Women's 1,000 Meters  China  Women   
2830  2014  Short-Track Speedskating      Men's 500 Meters  China    Men   

      Medal Rank Medalla Nombre_atleta_o_Equipo  Edad_Atleta  
1531           3  bronze                Chen Lu         17.0  
1732           2  silver              Xu Nannan         19.0  
1756           2  silver              An Yulong         19.0  
1765           2  silver          Yang Yang (S)         20.0  
1768           2  silver          Yang Yang (S)         20.0  
2229           1    gold              Wang Meng         20.0  
2233           2  silver              Wang Meng         20.0  
2237           3  bronze              Wang Meng         20.0  
2493           1    gold              Zhou Yang         18.0  
2616           2  silver             Tianyu Han         17.0  
2827           2  silver              Kexin Fan         20.0  
2830           2  silver              Dajing Wu         19.0

In [52]:
df[mask & mask_1 & mask_2]["Género"].value_counts()

Women    9
Men      3
Name: Género, dtype: int64

In [53]:
df[mask & mask_1 & mask_2]["Género"].value_counts(1)*100

Women    75.0
Men      25.0
Name: Género, dtype: float64

In [54]:
df[mask|mask_1|mask_2]["Género"].value_counts()

Men      1513
Women    1030
Mixed      94
Name: Género, dtype: int64


## _NaN_

Suele ocurrir que en nuestro `dataframe` se tengan valores perdidos. Estos deben ser tratados por métodos de imputación que mas adelante discutiremos. Por ahora veremos como eliminarlos o modificarlos.

In [55]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2865 entries, 0 to 2864
Data columns (total 9 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Año                     2865 non-null   int64  
 1   Deporte                 2865 non-null   object 
 2   Evento                  2865 non-null   object 
 3   País                    2865 non-null   object 
 4   Género                  2865 non-null   object 
 5   Medal Rank              2865 non-null   int64  
 6   Medalla                 2865 non-null   object 
 7   Nombre_atleta_o_Equipo  2865 non-null   object 
 8   Edad_Atleta             2173 non-null   float64
dtypes: float64(1), int64(2), object(6)
memory usage: 201.6+ KB


In [56]:
df_drop = df.dropna()
df_drop.head()

Año               Deporte               Evento     País Género  \
3  1924  Cross-Country Skiing  Men's 18 Kilometers   Norway    Men   
4  1924  Cross-Country Skiing  Men's 18 Kilometers   Norway    Men   
5  1924  Cross-Country Skiing  Men's 18 Kilometers  Finland    Men   
6  1924  Cross-Country Skiing  Men's 50 Kilometers   Norway    Men   
7  1924  Cross-Country Skiing  Men's 50 Kilometers   Norway    Men   

   Medal Rank Medalla  Nombre_atleta_o_Equipo  Edad_Atleta  
3           1    gold           Thorleif Haug         29.0  
4           2  silver  Johan Grøttumsbraaten         24.0  
5           3  bronze             Tapani Niku         28.0  
6           1    gold           Thorleif Haug         29.0  
7           2  silver      Thoralf Strømstad         27.0

In [57]:
df_drop.shape

(2173, 9)

In [58]:
df['Edad_Atleta'].isna()

0        True
1        True
2        True
3       False
4       False
        ...  
2860    False
2861    False
2862    False
2863     True
2864     True
Name: Edad_Atleta, Length: 2865, dtype: bool

In [59]:
df[df['Edad_Atleta'].isna()]

Año     Deporte            Evento           País Género  Medal Rank  \
0     1924     Bobsled   Men's Four/Five    Switzerland    Men           1   
1     1924     Bobsled   Men's Four/Five        Britain    Men           2   
2     1924     Bobsled   Men's Four/Five        Belgium    Men           3   
9     1924     Curling     Men's Curling        Britain    Men           1   
10    1924     Curling     Men's Curling         Sweden    Men           2   
...    ...         ...               ...            ...    ...         ...   
2857  2014     Bobsled        Men's Four         Russia    Men           1   
2858  2014     Bobsled        Men's Four         Latvia    Men           2   
2859  2014     Bobsled        Men's Four  United States    Men           3   
2863  2014  Ice Hockey  Men's Ice Hockey         Canada    Men           1   
2864  2014  Ice Hockey  Men's Ice Hockey         Sweden    Men           2   

     Medalla Nombre_atleta_o_Equipo  Edad_Atleta  
0       gold          Switzerland-1          NaN  
1     silver              Britain-1          NaN  
2     bronze              Belgium-1          NaN  
9       gold                Britain          NaN  
10    silver                 Sweden          NaN  
...      ...                    ...          ...  
2857    gold                 Russia          NaN  
2858  silver                 Latvia          NaN  
2859  bronze          United States          NaN  
2863    gold                 Canada          NaN  
2864  silver                 Sweden          NaN  

[692 rows x 9 columns]

In [60]:
## Podemos obtener la negación de la máscara
df[~(df['Edad_Atleta'].isna())]

Año               Deporte                 Evento      País Género  \
3     1924  Cross-Country Skiing    Men's 18 Kilometers    Norway    Men   
4     1924  Cross-Country Skiing    Men's 18 Kilometers    Norway    Men   
5     1924  Cross-Country Skiing    Men's 18 Kilometers   Finland    Men   
6     1924  Cross-Country Skiing    Men's 50 Kilometers    Norway    Men   
7     1924  Cross-Country Skiing    Men's 50 Kilometers    Norway    Men   
...    ...                   ...                    ...       ...    ...   
2849  2014          Snowboarding  Men's Parallel Slalom  Slovenia    Men   
2850  2014          Snowboarding  Men's Parallel Slalom   Austria    Men   
2860  2014  Cross-Country Skiing    Men's 50 Kilometers    Russia    Men   
2861  2014  Cross-Country Skiing    Men's 50 Kilometers    Russia    Men   
2862  2014  Cross-Country Skiing    Men's 50 Kilometers    Russia    Men   

      Medal Rank Medalla  Nombre_atleta_o_Equipo  Edad_Atleta  
3              1    gold           Thorleif Haug         29.0  
4              2  silver  Johan Grøttumsbraaten         24.0  
5              3  bronze             Tapani Niku         28.0  
6              1    gold           Thorleif Haug         29.0  
7              2  silver      Thoralf Strømstad         27.0  
...          ...     ...                     ...          ...  
2849           2  silver               Zan Kosir         29.0  
2850           3  bronze           Benjamin Karl         28.0  
2860           1    gold        Alexander Legkov         30.0  
2861           2  silver       Maxim Vylegzhanin         31.0  
2862           3  bronze         Ilia Chernousov         27.0  

[2173 rows x 9 columns]

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2865 entries, 0 to 2864
Data columns (total 9 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Año                     2865 non-null   int64  
 1   Deporte                 2865 non-null   object 
 2   Evento                  2865 non-null   object 
 3   País                    2865 non-null   object 
 4   Género                  2865 non-null   object 
 5   Medal Rank              2865 non-null   int64  
 6   Medalla                 2865 non-null   object 
 7   Nombre_atleta_o_Equipo  2865 non-null   object 
 8   Edad_Atleta             2173 non-null   float64
dtypes: float64(1), int64(2), object(6)
memory usage: 201.6+ KB


In [62]:
## Todo regístro None o NaN es reemplazado por 0
df.fillna("Alerta!!!")

Año               Deporte               Evento         País Género  \
0     1924               Bobsled      Men's Four/Five  Switzerland    Men   
1     1924               Bobsled      Men's Four/Five      Britain    Men   
2     1924               Bobsled      Men's Four/Five      Belgium    Men   
3     1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4     1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
...    ...                   ...                  ...          ...    ...   
2860  2014  Cross-Country Skiing  Men's 50 Kilometers       Russia    Men   
2861  2014  Cross-Country Skiing  Men's 50 Kilometers       Russia    Men   
2862  2014  Cross-Country Skiing  Men's 50 Kilometers       Russia    Men   
2863  2014            Ice Hockey     Men's Ice Hockey       Canada    Men   
2864  2014            Ice Hockey     Men's Ice Hockey       Sweden    Men   

      Medal Rank Medalla  Nombre_atleta_o_Equipo Edad_Atleta  
0              1    gold           Switzerland-1   Alerta!!!  
1              2  silver               Britain-1   Alerta!!!  
2              3  bronze               Belgium-1   Alerta!!!  
3              1    gold           Thorleif Haug        29.0  
4              2  silver  Johan Grøttumsbraaten        24.0  
...          ...     ...                     ...         ...  
2860           1    gold        Alexander Legkov        30.0  
2861           2  silver       Maxim Vylegzhanin        31.0  
2862           3  bronze         Ilia Chernousov        27.0  
2863           1    gold                  Canada   Alerta!!!  
2864           2  silver                  Sweden   Alerta!!!  

[2865 rows x 9 columns]

In [63]:
df["Edad_Atleta"].fillna(0)

0        0.0
1        0.0
2        0.0
3       29.0
4       24.0
        ... 
2860    30.0
2861    31.0
2862    27.0
2863     0.0
2864     0.0
Name: Edad_Atleta, Length: 2865, dtype: float64

In [64]:
df["Edad_Atleta_imputado"] = df["Edad_Atleta"].fillna(df["Edad_Atleta"].mean())
df.head(10)

Año               Deporte               Evento         País Género  \
0  1924               Bobsled      Men's Four/Five  Switzerland    Men   
1  1924               Bobsled      Men's Four/Five      Britain    Men   
2  1924               Bobsled      Men's Four/Five      Belgium    Men   
3  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
5  1924  Cross-Country Skiing  Men's 18 Kilometers      Finland    Men   
6  1924  Cross-Country Skiing  Men's 50 Kilometers       Norway    Men   
7  1924  Cross-Country Skiing  Men's 50 Kilometers       Norway    Men   
8  1924  Cross-Country Skiing  Men's 50 Kilometers       Norway    Men   
9  1924               Curling        Men's Curling      Britain    Men   

   Medal Rank Medalla  Nombre_atleta_o_Equipo  Edad_Atleta  \
0           1    gold           Switzerland-1          NaN   
1           2  silver               Britain-1          NaN   
2           3  bronze               Belgium-1          NaN   
3           1    gold           Thorleif Haug         29.0   
4           2  silver  Johan Grøttumsbraaten         24.0   
5           3  bronze             Tapani Niku         28.0   
6           1    gold           Thorleif Haug         29.0   
7           2  silver      Thoralf Strømstad         27.0   
8           3  bronze  Johan Grøttumsbraaten         24.0   
9           1    gold                 Britain          NaN   

   Edad_Atleta_imputado  
0             25.152324  
1             25.152324  
2             25.152324  
3             29.000000  
4             24.000000  
5             28.000000  
6             29.000000  
7             27.000000  
8             24.000000  
9             25.152324

In [65]:
df['Edad_Atleta'].value_counts(dropna = False)

NaN     692
24.0    212
23.0    200
27.0    193
25.0    178
28.0    176
26.0    168
22.0    166
21.0    146
20.0    125
29.0    117
30.0     85
19.0     85
31.0     62
32.0     57
33.0     42
18.0     41
34.0     29
17.0     27
35.0     13
16.0     12
15.0      8
36.0      8
37.0      7
38.0      6
41.0      4
39.0      2
40.0      2
14.0      1
42.0      1
Name: Edad_Atleta, dtype: int64

In [66]:
df.describe().T

count         mean        std     min     25%  \
Año                   2865.0  1986.488656  23.489775  1924.0  1972.0   
Medal Rank            2865.0     1.995812   0.815773     1.0     1.0   
Edad_Atleta           2173.0    25.152324   4.271764    14.0    22.0   
Edad_Atleta_imputado  2865.0    25.152324   3.720066    14.0    23.0   

                              50%     75%     max  
Año                   1992.000000  2006.0  2014.0  
Medal Rank               2.000000     3.0     3.0  
Edad_Atleta             25.000000    28.0    42.0  
Edad_Atleta_imputado    25.152324    27.0    42.0

In [67]:
df_drop.shape

(2173, 9)

## `pandas.crosstab`

La función `pandas.crosstab` devuelve la tabla de contingencia resultante de cruzar dos o más columnas de un dataframe. Por defecto ell resultado evalúa las frecuencias (absolutas o relativas) de cada combinación de valores, es posible especificar una función de agregación.

In [68]:
pd.crosstab(df['Género'],df['Medalla'])

Medalla  bronze  gold  silver
Género                       
Men         568   573     570
Mixed        36    37      36
Women       344   349     352

In [69]:
## Podemos obtener la suma por filas y columnas (subtotales) y renombrar esta
pd.crosstab(df['Género'],df['Medalla'], margins=True, margins_name="Total",)

Medalla  bronze  gold  silver  Total
Género                              
Men         568   573     570   1711
Mixed        36    37      36    109
Women       344   349     352   1045
Total       948   959     958   2865

In [70]:
## Podemos mostrar las frecuencias "normalizadas" (es decir una parte de la unidad). Se puede obtener respecto 
## al total o con respecto a los subtotales de filas o columnas. Asignamos al parámetro normalize el valor True,
## la normalización se realiza con respecto al total de muestras
pd.crosstab(df['Género'],df['Medalla'], normalize=True)

Medalla    bronze      gold    silver
Género                               
Men      0.198255  0.200000  0.198953
Mixed    0.012565  0.012914  0.012565
Women    0.120070  0.121815  0.122862

Del total de las medallas entregadas: 

1.   19.8% fueron de bronce y entregas a hombres.
2.   12.2% fueron de plata y entregadas a mujeres.

En cada casilla podemos extraer información similar.


In [71]:
pd.crosstab(df['Género'],df['Medalla'], normalize=True, margins=True, margins_name= "Total %")

Medalla    bronze      gold    silver   Total %
Género                                         
Men      0.198255  0.200000  0.198953  0.597208
Mixed    0.012565  0.012914  0.012565  0.038045
Women    0.120070  0.121815  0.122862  0.364747
Total %  0.330890  0.334729  0.334380  1.000000

In [72]:
pd.crosstab(df['Género'],df['País'])

País    Australia  Austria  Belarus  Belgium  Britain  Bulgaria  Canada  \
Género                                                                    
Men             5      143        8        4       13         1      82   
Mixed           0        5        0        1        2         0      11   
Women           7       70        7        0       11         5      77   

País    China  Croatia  Czech Republic  ...  Soviet Union  Spain  Sweden  \
Género                                  ...                                
Men        12        5               9  ...           104      1     106   
Mixed       5        0               1  ...            20      0       0   
Women      36        6              14  ...            70      1      38   

País    Switzerland  Ukraine  Unified Team  United States  Uzbekistan  \
Género                                                                  
Men              98        0             4            155           0   
Mixed             0        1             4             11           0   
Women            40        6            15            116           1   

País    West Germany  Yugoslavia  
Género                            
Men               25           3  
Mixed              1           0  
Women             13           1  

[3 rows x 44 columns]

In [73]:
pais_genero = pd.DataFrame(pd.crosstab(df['Género'],df['País']).unstack()).rename({0:"Conteo"},axis=1)
pais_genero.head(30)

Conteo
País           Género        
Australia      Men          5
               Mixed        0
               Women        7
Austria        Men        143
               Mixed        5
               Women       70
Belarus        Men          8
               Mixed        0
               Women        7
Belgium        Men          4
               Mixed        1
               Women        0
Britain        Men         13
               Mixed        2
               Women       11
Bulgaria       Men          1
               Mixed        0
               Women        5
Canada         Men         82
               Mixed       11
               Women       77
China          Men         12
               Mixed        5
               Women       36
Croatia        Men          5
               Mixed        0
               Women        6
Czech Republic Men          9
               Mixed        1
               Women       14

In [74]:
pais_genero.index

MultiIndex([(    'Australia',   'Men'),
            (    'Australia', 'Mixed'),
            (    'Australia', 'Women'),
            (      'Austria',   'Men'),
            (      'Austria', 'Mixed'),
            (      'Austria', 'Women'),
            (      'Belarus',   'Men'),
            (      'Belarus', 'Mixed'),
            (      'Belarus', 'Women'),
            (      'Belgium',   'Men'),
            ...
            ('United States', 'Women'),
            (   'Uzbekistan',   'Men'),
            (   'Uzbekistan', 'Mixed'),
            (   'Uzbekistan', 'Women'),
            ( 'West Germany',   'Men'),
            ( 'West Germany', 'Mixed'),
            ( 'West Germany', 'Women'),
            (   'Yugoslavia',   'Men'),
            (   'Yugoslavia', 'Mixed'),
            (   'Yugoslavia', 'Women')],
           names=['País', 'Género'], length=132)

In [75]:
pais_genero.loc["Belgium"]

Conteo
Género        
Men          4
Mixed        1
Women        0

## _Conclusiones._

Revisamos el método de importación de archivos en distintios formatos. Además, se vieron varios métodos de consulta de los `dataframe`. Se revisó `pandas.croostab` es una tabla de contingencía generada apartir del `dataframe.`